File: sdk/node/src/api/api_ws.ts
import { SpotPublicWS } from '@generate/spot/spotpublic';
import { SpotPrivateWS } from '@generate/spot/spotprivate';
import { MarginPublicWS } from '@generate/margin/marginpublic';
import { MarginPrivateWS } from '@generate/margin/marginprivate';
import { FuturesPublicWS } from '@generate/futures/futurespublic';
import { FuturesPrivateWS } from '@generate/futures/futuresprivate';
import { UtaPublicWS } from '@generate/uta/publicws';
import { UtaPrivateTradeWS, UtaPrivateWS } from '@generate/uta/privatews';
import { PushTradeType } from '@model/push_trade_type';

export interface KucoinWSService {
    /**
     * Returns the interface to interact with the Spot Trading WebSocket (public channel) API of KuCoin.
     */
    newSpotPublicWS(): SpotPublicWS;

    /**
     * Returns the interface to interact with the Spot Trading WebSocket (private channel) API of KuCoin.
     */
    newSpotPrivateWS(): SpotPrivateWS;

    /**
     * Returns the interface to interact with the Margin Trading WebSocket (public channel) API of KuCoin.
     */
    newMarginPublicWS(): MarginPublicWS;

    /**
     * Returns the interface to interact with the Margin Trading WebSocket (private channel) API of KuCoin.
     */
    newMarginPrivateWS(): MarginPrivateWS;

    /**
     * Returns the interface to interact with the Futures Trading WebSocket (public channel) API of KuCoin.
     */
    newFuturesPublicWS(): FuturesPublicWS;

    /**
     * Returns the interface to interact with the Futures Trading WebSocket (private channel) API of KuCoin.
     */
    newFuturesPrivateWS(): FuturesPrivateWS;

    /** Returns a direct UTA public push WebSocket for Spot or Futures. */
    newUtaPublicWS(tradeType: PushTradeType): UtaPublicWS;

    /** Returns the authenticated direct UTA private push WebSocket. */
    newUtaPrivateWS(): UtaPrivateWS;

    /** Returns the authenticated direct UTA trading WebSocket. */
    newUtaPrivateTradeWS(): UtaPrivateTradeWS;
}

File: sdk/golang/pkg/api/api_ws.go
package api

import (
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/futures/futuresprivate"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/futures/futurespublic"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/margin/marginprivate"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/margin/marginpublic"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/spot/spotprivate"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/spot/spotpublic"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/uta/privatews"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/uta/publicws"
)

type KucoinWSService interface {
	// NewSpotPublicWS returns the interface to interact with
	// the Spot Trading websocket(public channel) API of Kucoin.
	NewSpotPublicWS() spotpublic.SpotPublicWS

	// NewSpotPrivateWS returns the interface to interact with
	// the Spot Trading websocket(private channel) API of Kucoin.
	NewSpotPrivateWS() spotprivate.SpotPrivateWS

	// NewMarginPublicWS returns the interface to interact with
	// the Margin Trading websocket(public channel) API of Kucoin.
	NewMarginPublicWS() marginpublic.MarginPublicWS

	// NewMarginPrivateWS returns the interface to interact with
	// the Margin Trading websocket(private channel) API of Kucoin.
	NewMarginPrivateWS() marginprivate.MarginPrivateWS

	// NewFuturesPublicWS returns the interface to interact with
	// the Futures Trading websocket(public channel) API of Kucoin.
	NewFuturesPublicWS() futurespublic.FuturesPublicWS

	// NewFuturesPrivateWS returns the interface to interact with
	// the Futures Trading websocket(private channel) API of Kucoin.
	NewFuturesPrivateWS() futuresprivate.FuturesPrivateWS

	// NewUtaPublicWS returns a direct UTA public push WebSocket service for SPOT or FUTURES.
	NewUtaPublicWS(tradeType publicws.PushTradeType) publicws.UtaPublicWS

	// NewUtaPrivateWS returns the authenticated direct UTA private push WebSocket service.
	NewUtaPrivateWS() privatews.UtaPrivateWS

	// NewUtaPrivateTradeWS returns the authenticated direct UTA WebSocket trading service.
	NewUtaPrivateTradeWS() privatews.UtaPrivateTradeWS
}
